File: agents/selected_template.py
"""Immutable, compiled selected-template constraints for both search engines.

Coordinates are zero based, bottom up. Colors are compact IDs 1..5; forbidden
color 0 means any occupied cell. Unlisted cells are unconstrained. No catalog
matching or rebinding occurs here. Satisfied required cells must survive every
subsequent resolved placement, including after completion.
"""

from __future__ import annotations

from dataclasses import dataclass
import hashlib
import struct
import time

from agents.compact_search import CompactSearchState

SCHEMA = "puyo.selected_template.v1"
ABI_VERSION = 1


@dataclass(frozen=True, slots=True)
class SelectedTemplate:
    template_id: str
    variant_id: str
    transform: str
    binding: tuple[tuple[str, int], ...]
    required_cells: tuple[tuple[int, int, int], ...]
    forbidden_cells: tuple[tuple[int, int, int], ...] = ()
    schema_version: str = SCHEMA

    def __post_init__(self):
        if self.schema_version != SCHEMA:
            raise ValueError("unsupported selected-template schema")
        for name in (self.template_id, self.variant_id, self.transform):
            if not isinstance(name, str) or not name or len(name.encode()) > 4096:
                raise ValueError("selected-template identity must be a bounded string")
        binding = tuple(sorted(tuple(item) for item in self.binding))
        if not 1 <= len(binding) <= 5 or len({v[0] for v in binding}) != len(binding):
            raise ValueError("selected-template binding requires unique labels")
        if len({v[1] for v in binding}) != len(binding):
            raise ValueError("selected-template binding must be injective")
        for label, color in binding:
            if (
                not isinstance(label, str)
                or not label
                or len(label.encode()) > 4096
                or type(color) is not int
                or not 1 <= color <= 5
            ):
                raise ValueError("invalid selected-template binding")
        object.__setattr__(self, "binding", binding)
        colors = {c for _, c in binding}
        for name in ("required_cells", "forbidden_cells"):
            cells = tuple(sorted(tuple(cell) for cell in getattr(self, name)))
            if len(cells) > 84 * 6 or len(set(cells)) != len(cells):
                raise ValueError("duplicate or excessive selected-template cells")
            for x, y, color in cells:
                if (
                    any(type(v) is not int for v in (x, y, color))
                    or not 0 <= x < 6
                    or not 0 <= y < 14
                    or color
                    not in colors | ({0} if name == "forbidden_cells" else set())
                ):
                    raise ValueError("invalid selected-template cell")
            if name == "required_cells" and (
                not cells or len({(x, y) for x, y, _ in cells}) != len(cells)
            ):
                raise ValueError("required cells must have unique coordinates")
            object.__setattr__(self, name, cells)
        for x, y, color in self.required_cells:
            if (x, y, color) in self.forbidden_cells or (
                x,
                y,
                0,
            ) in self.forbidden_cells:
                raise ValueError("contradictory selected-template cells")

    def to_bytes(self) -> bytes:
        result = bytearray(struct.pack("<H", ABI_VERSION))
        for value in (
            self.schema_version,
            self.template_id,
            self.variant_id,
            self.transform,
        ):
            encoded = value.encode()
            result.extend(struct.pack("<H", len(encoded)) + encoded)
        result.append(len(self.binding))
        for label, color in self.binding:
            encoded = label.encode()
            result.extend(struct.pack("<H", len(encoded)) + encoded + bytes([color]))
        for cells in (self.required_cells, self.forbidden_cells):
            result.extend(struct.pack("<H", len(cells)))
            for cell in cells:
                result.extend(bytes(cell))
        return bytes(result)

    @property
    def semantic_digest(self) -> str:
        return hashlib.sha256(self.to_bytes()).hexdigest()

    @classmethod
    def from_bytes(cls, payload: bytes):
        from agents.deep_chain_native import _Reader, IncompatibleSchemaError

        r = _Reader(payload, failing_tag=0x8007)
        if r.u16("selected-template ABI") != ABI_VERSION:
            raise IncompatibleSchemaError("unsupported selected-template ABI")
        schema, identity, variant, transform = (
            r.string("template identity") for _ in range(4)
        )
        if schema != SCHEMA:
            raise IncompatibleSchemaError("unsupported selected-template schema")
        binding = tuple(
            (r.string("binding label"), r.u8("binding color"))
            for _ in range(r.u8("binding count"))
        )
        required, forbidden = (
            tuple(
                tuple(r.u8("cell") for _ in range(3))
                for _ in range(r.u16("cell count"))
            )
            for _ in range(2)
        )
        r.finish()
        value = cls(identity, variant, transform, binding, required, forbidden, schema)
        if value.to_bytes() != payload:
            raise ValueError("selected-template encoding is not canonical")
        return value

    def evaluate(
        self, state: CompactSearchState, parent: CompactSearchState
    ) -> tuple[bool, bool]:
        complete = True
        for x, y, color in self.required_cells:
            bit = 1 << (y * 6 + x)
            matches = bool(state.planes[color - 1] & bit)
            if not matches:
                complete = False
                if state.occupied_mask & bit or parent.planes[color - 1] & bit:
                    return False, False
        for x, y, color in self.forbidden_cells:
            plane = state.occupied_mask if color == 0 else state.planes[color - 1]
            if plane & (1 << (y * 6 + x)):
                return False, False
        return True, complete


def validate_public_template_input(template, known_pairs, config):
    if template is not None:
        if not isinstance(template, SelectedTemplate):
            raise ValueError("selected_template must be an immutable SelectedTemplate")
        if not 1 <= len(known_pairs) <= 3:
            raise ValueError("selected-template accepts public current/NEXT/NEXT2 only")
        if config.fire_context != "safe_build":
            raise ValueError(
                "selected-template must not constrain the survival safety path"
            )


def new_template_record():
    return {
        "checks": 0,
        "rejected": 0,
        "_check_ns": 0,
        "root_violation": False,
        "known_witness": (),
        "sampled_witness": (),
        "sampled_scenario_id": None,
    }


def check_template(
    template,
    records,
    action,
    state,
    parent,
    path,
    known_count,
    initial_valid,
    scenario_id,
):
    record = records[action]
    record["checks"] += 1
    started = time.perf_counter_ns()
    valid, complete = template.evaluate(state, parent)
    record["_check_ns"] += time.perf_counter_ns() - started
    if not initial_valid or not valid:
        record["rejected"] += 1
        if len(path) == 1:
            record["root_violation"] = True
        return False
    if complete and not state.game_over:
        name = "known_witness" if len(path) <= known_count else "sampled_witness"
        previous = record[name]
        if (
            not previous
            or (len(path), path) < (len(previous), previous)
            or (
                name == "sampled_witness"
                and path == previous
                and scenario_id < record["sampled_scenario_id"]
            )
        ):
            record[name] = path
            if name == "sampled_witness":
                record["sampled_scenario_id"] = scenario_id
    return True


def template_result(template, records, counters, representatives):
    if template is None:
        return None
    roots = {}
    for action, value in sorted(records.items()):
        record = {key: item for key, item in value.items() if not key.startswith("_")}
        record["status"] = (
            "violated"
            if record["root_violation"]
            else "known_complete"
            if record["known_witness"]
            else "cutoff"
            if counters.budget_exhausted
            else "unknown"
        )
        record["reason"] = (
            "resolved_constraint_violation"
            if record["root_violation"]
            else "public_prefix_completion"
            if record["known_witness"]
            else "request_node_quota"
            if counters.budget_exhausted
            else "sampled_completion_only"
            if record["sampled_witness"]
            else "no_completion_witness"
        )
        record["known_scenario_id"] = None  # Public prefix is scenario-independent.
        record["completion_source"] = (
            "public"
            if record["known_witness"]
            else "sampled"
            if record["sampled_witness"]
            else None
        )
        record["compatible"] = (
            action in representatives and not record["root_violation"]
        )
        roots[action] = record
    return {
        "schema_version": SCHEMA,
        "semantic_digest": template.semantic_digest,
        "roots": roots,
    }


def decode_template_result(
    payload,
    template,
    counters,
    representatives,
    legal_roots,
    known_count,
    depth,
    legal_scenarios,
):
    from agents.deep_chain_native import _Reader, InvalidNativeInputError

    if template is None:
        if payload is not None:
            raise InvalidNativeInputError("unexpected selected-template result")
        return None
    if payload is None:
        raise InvalidNativeInputError("missing selected-template result")
    r = _Reader(payload, failing_tag=0x8307)
    if (
        r.u16("template result ABI") != ABI_VERSION
        or r.take(32, "template digest").hex() != template.semantic_digest
    ):
        raise InvalidNativeInputError("selected-template result identity mismatch")
    records = {}
    for _ in range(r.u16("template root count")):
        action = r.u8("template root action")
        checks, rejected = r.u64("checks"), r.u64("rejected")
        violation = r.u8("root violation")
        known, sampled = (
            tuple(r.u8("witness action") for _ in range(r.u8("witness length")))
            for _ in range(2)
        )
        sampled_scenario = r.u8("sampled witness scenario")
        if (
            (not sampled and sampled_scenario != 255)
            or (sampled and sampled_scenario not in legal_scenarios)
            or action in records
            or action not in legal_roots
            or rejected > checks
            or violation > 1
            or (violation and (not rejected or known or sampled))
            or any(a >= 22 for a in known + sampled)
            or any(path and path[0] != action for path in (known, sampled))
            or len(known) > min(known_count, depth)
            or (sampled and not known_count < len(sampled) <= depth)
        ):
            raise InvalidNativeInputError("invalid selected-template root evidence")
        records[action] = dict(
            checks=checks,
            rejected=rejected,
            root_violation=bool(violation),
            known_witness=known,
            sampled_witness=sampled,
            sampled_scenario_id=None if sampled_scenario == 255 else sampled_scenario,
        )
    r.u64("template check nanoseconds")
    r.finish()
    if (
        set(records) != set(legal_roots)
        or sum(v["checks"] for v in records.values()) != counters.generated_nodes
    ):
        raise InvalidNativeInputError("incomplete selected-template accounting")
    return template_result(template, records, counters, representatives)

File: native/deep_chain_native/src/selected_template.rs
//! Versioned, request-local fixed binding. Unlisted cells are unconstrained.
use crate::compact::CompactState;
use crate::{ContractError, ContractResult, Reader};
use sha2::{Digest, Sha256};
use std::time::Instant;

pub(crate) const TAG: u16 = 0x8007;
pub(crate) const RESULT_TAG: u16 = 0x8307;

#[derive(Clone)]
pub(crate) struct SelectedTemplate {
    pub(crate) digest: [u8; 32],
    required: [u128; 5],
    forbidden: [u128; 6],
}
impl SelectedTemplate {
    pub(crate) fn parse(data: &[u8]) -> ContractResult<Self> {
        let mut r = Reader::new(data, TAG);
        if r.u16("template ABI")? != 1
            || r.string("template schema")? != "puyo.selected_template.v1"
        {
            return Err(ContractError::incompatible(
                TAG,
                "unsupported selected-template schema/ABI",
            ));
        }
        for _ in 0..3 {
            if r.string("template identity")?.is_empty() {
                return Err(ContractError::invalid(TAG, "empty template identity"));
            }
        }
        let count = r.u8("binding count")?;
        if count == 0 || count > 5 {
            return Err(ContractError::invalid(TAG, "invalid binding count"));
        }
        let mut colors = 0_u8;
        let mut last_label = String::new();
        for _ in 0..count {
            let label = r.string("binding label")?;
            let color = r.u8("binding color")?;
            if label <= last_label || !(1..=5).contains(&color) || colors & (1 << color) != 0 {
                return Err(ContractError::invalid(
                    TAG,
                    "invalid canonical fixed binding",
                ));
            }
            last_label = label;
            colors |= 1 << color;
        }
        let mut required = [0_u128; 5];
        let mut forbidden = [0_u128; 6];
        let mut required_mask = 0_u128;
        for kind in 0..2 {
            let count = r.u16("cell count")?;
            if count > 504 || (kind == 0 && count == 0) {
                return Err(ContractError::invalid(TAG, "invalid template cell count"));
            }
            let mut previous = None;
            for _ in 0..count {
                let x = r.u8("cell x")?;
                let y = r.u8("cell y")?;
                let color = r.u8("cell color")?;
                if x >= 6
                    || y >= 14
                    || color > 5
                    || (color == 0 && kind == 0)
                    || (color != 0 && colors & (1 << color) == 0)
                    || previous.is_some_and(|p| p >= (x, y, color))
                {
                    return Err(ContractError::invalid(
                        TAG,
                        "invalid canonical template cell",
                    ));
                }
                previous = Some((x, y, color));
                let bit = 1_u128 << (y * 6 + x);
                if kind == 0 {
                    if required_mask & bit != 0 {
                        return Err(ContractError::invalid(TAG, "duplicate required coordinate"));
                    }
                    required_mask |= bit;
                    required[usize::from(color - 1)] |= bit;
                } else {
                    if (color == 0 && required_mask & bit != 0)
                        || (color > 0 && required[usize::from(color - 1)] & bit != 0)
                    {
                        return Err(ContractError::invalid(TAG, "contradictory template cells"));
                    }
                    forbidden[usize::from(color)] |= bit;
                }
            }
        }
        r.finish()?;
        Ok(Self {
            digest: Sha256::digest(data).into(),
            required,
            forbidden,
        })
    }

    pub(crate) fn evaluate(&self, state: &CompactState, parent: &CompactState) -> (bool, bool) {
        let planes = state.wire_planes();
        let parent_planes = parent.wire_planes();
        let occupied = state.occupied();
        let mut complete = true;
        for i in 0..5 {
            let missing = self.required[i] & !planes[i];
            if missing != 0 {
                complete = false;
            }
            if missing & (occupied | parent_planes[i]) != 0
                || planes[i] & self.forbidden[i + 1] != 0
            {
                return (false, false);
            }
        }
        if occupied & self.forbidden[0] != 0 {
            return (false, false);
        }
        (true, complete)
    }
}

#[derive(Clone, Debug, Default)]
pub(crate) struct Record {
    pub(crate) checks: u64,
    pub(crate) check_ns: u64,
    pub(crate) rejected: u64,
    pub(crate) root_violation: bool,
    pub(crate) known: Vec<u8>,
    pub(crate) sampled: Vec<u8>,
    pub(crate) sampled_scenario: Option<u8>,
}
impl Record {
    pub(crate) fn check(
        &mut self,
        template: &SelectedTemplate,
        state: &CompactState,
        parent: &CompactState,
        path: &[u8],
        known_count: usize,
        initial_valid: bool,
        game_over: bool,
        scenario_id: u8,
    ) -> bool {
        self.checks += 1;
        let started = Instant::now();
        let (valid, complete) = template.evaluate(state, parent);
        self.check_ns += started.elapsed().as_nanos().min(u128::from(u64::MAX)) as u64;
        if !initial_valid || !valid {
            self.rejected += 1;
            self.root_violation |= path.len() == 1;
            return false;
        }
        if complete && !game_over {
            let witness = if path.len() <= known_count {
                &mut self.known
            } else {
                &mut self.sampled
            };
            if witness.is_empty() || (path.len(), path) < (witness.len(), witness.as_slice()) {
                *witness = path.to_vec();
                if path.len() > known_count {
                    self.sampled_scenario = Some(scenario_id);
                }
            }
        }
        true
    }
    pub(crate) fn add(&mut self, other: &Self) {
        self.checks += other.checks;
        self.check_ns += other.check_ns;
        self.rejected += other.rejected;
        self.root_violation |= other.root_violation;
        if !other.known.is_empty()
            && (self.known.is_empty()
                || (other.known.len(), &other.known) < (self.known.len(), &self.known))
        {
            self.known = other.known.clone();
        }
        if !other.sampled.is_empty()
            && (self.sampled.is_empty()
                || (other.sampled.len(), &other.sampled, other.sampled_scenario)
                    < (self.sampled.len(), &self.sampled, self.sampled_scenario))
        {
            self.sampled = other.sampled.clone();
            self.sampled_scenario = other.sampled_scenario;
        }
    }
}
